server/utils.go

```diff
@@ -424,42 +390,76 @@ func versionToString(vers int) string {
 	return str
 }
 
-// rewriteToken attempts to match the original token
-// against the email, telephone number or login patterns.
-// On success, prepends the token with the corresponding prefix.
-func rewriteToken(orig, countryCode string) string {
-	if orig == "" || prefixedTagRegexp.MatchString(orig) {
-		// It's either empty or already has a prefix e.g. basic:alice.
+// Tag handling
+
+// Take a slice of tags, return a slice of restricted namespace tags contained in the input.
+// Tags to filter, restricted namespaces to filter.
+func filterRestrictedTags(tags []string, namespaces map[string]bool) []string {
+	var out []string
+	if len(namespaces) == 0 {
+		return out
+	}
+
+	for _, s := range tags {
+		parts := prefixedTagRegexp.FindStringSubmatch(s)
+
+		if len(parts) < 2 {
+			continue
+		}
+
+		if namespaces[parts[1]] {
+			out = append(out, s)
+		}
+	}
+
+	return out
+}
+
+// rewriteToken attempts to match the original token against the email, telephone number and optionally login patterns.
+// The tag is expected to be converted to lowercase.
+// On success, it prepends the token with the corresponding prefix. It returns an empty string if the tag is invalid.
+// TODO: better handling of countryCode:
+// 1. As provided by the client (e.g. as specified or inferred from client's phone number or location).
+// 2. Use value from the .conf file.
+// 3. Fallback to US as a last resort.
+func rewriteTag(orig, countryCode string, withLogin bool) string {
+	// Check if the tag already has a prefix e.g. basic:alice.
+	if prefixedTagRegexp.MatchString(orig) {
 		return orig
 	}
+
 	// Is it email?
-	if r, err := mail.ParseAddress(orig); err == nil && r.Address == orig {
-		return "email:" + orig
-	}
-	// TODO: pass region as a param.
-	// 1. As provided by the client (e.g. as specified or inferred
-	//    from client's phone number or location).
-	// 2. Use value from .conf file.
-	// 3. Fallback to US as a last resort.
+	if addr, err := mail.ParseAddress(orig); err == nil {
+		if len([]rune(addr.Address)) < maxTagLength && addr.Address == orig {
+			return "email:" + orig
+		}
+		return ""
+	}
+
 	if num, err := phonenumbers.Parse(orig, countryCode); err == nil {
-		// It's a phone number.
+		// It's a phone number. Not checking the length because phone numbers cannot be that long.
 		return "tel:" + phonenumbers.Format(num, phonenumbers.E164)
 	}
+
 	// Does it look like a username/login?
 	// TODO: use configured authenticators to check if orig is a valid user name.
-	if basicLoginName.MatchString(orig) {
+	if withLogin && basicLoginName.MatchString(orig) {
 		return "basic:" + orig
 	}
-	return orig
+
+	if tagRegexp.MatchString(orig) {
+		return orig
+	}
+	return ""
```

Comment:
Log an error here as well?
It may be useful for us to know what kind of input we skip.

Reply:
I've added the logging, but it does not look that useful. This is a tag with spaces and such or too long tag.
I don't see an easy way to log just the unparsable phone numbers.